drivers/i2c/i2c_dw.c

```diff
@@ -42,7 +42,12 @@ LOG_MODULE_REGISTER(i2c_dw);
 
 static inline uint32_t get_regs(const struct device *dev)
 {
+#ifdef CONFIG_SOC_CYCLONE5  /* adding an exception for Cyclone V SoC FPGA SoC */
```

Comment:
why?

Reply:
This is an interesting one...
The approach used in this driver was to use the function DEVICE_MMIO_GET, which wasn't working for us.

This function was defined in include/zephyr/sys/device_mmio.h... And as you can see it calls DEVICE_MMIO_RAM_PTR
which is defined as (mm_reg_t *)((device)->data), hence, it doesn't return the register base address as we want.

I'm open to suggestions, but we determined that this was the most evident way to get this.

Reply:
Something is fishy here. Do you have `CONFIG_MMU` or `CONFIG_PCIE` enabled?

Reply:
CONFIG_MMU 1
in the generated autoconf.h

CONFIG_PCIE is not set

Reply:
Right, so if you have the MMU enabled that means that you definitely do not want to use the physical address to access the I2C because almost definitely `phys != virt`. So this is definitely wrong.

Reply:
but, if I dont use this, my i2c simply does not work, thats why I introduced the line... If I get the base address as I did, it works every time....
If I just use whats already in place (DEVICE_MMIO_GET) it simply does not work

Reply:
It is possible that your MMU is 1:1 mapping between phys and virt, but this is still wrong. Why `DEVICE_MMIO_GET` is not working? what is the address returned? Is your MMU correctly mapping the address?

We do not want to add quirks to workaround issues :)

Reply:
DEVICE_MMIO_GET returns 0
And yes, I was also doubtful about this fix, because it is very specific for the CycloneV... But when I started debugging, I saw that the code in device_mmio.h is not helping me at all ..., is either adding this line or in some way modify .h files in a similar way

Reply:
`DEVICE_MMIO_GET` should return the `virt` address mapped by the `DEVICE_MMIO_MAP` macro in `i2c_dw_initialize()`.  `DEVICE_MMIO_MAP` is calling `device_map()` to MMU-map the address. Now, what `virt` address is being returned by the `device_map()` call?

Reply:
Moving the get_regs() after DEVICE_MMIO_MAP seems to do the trick. I'll remove this file from this PR, and created another PR to fix that particular bug!